src/Storages/StorageMerge.cpp

```diff
@@ -107,13 +107,32 @@ StoragePtr StorageMerge::getFirstTable(F && predicate) const
     return {};
 }
 
+template <typename F>
+void StorageMerge::forEachTable(F && func) const
+{
+    getFirstTable([&func](const auto & table)
+    {
+        func(table);
+        return false;
+    });
+}
 
 bool StorageMerge::isRemote() const
 {
     auto first_remote_table = getFirstTable([](const StoragePtr & table) { return table && table->isRemote(); });
     return first_remote_table != nullptr;
 }
 
+bool StorageMerge::canMoveConditionsToPrewhere() const
+{
+    /// NOTE: This check is used during query analysis as condition for applying
+    /// "move to PREWHERE" optimization. However, it contains a logical race:
+    /// If new table that matches regexp for current storage and doesn't support PREWHERE
+    /// will appear after this check and before calling "read" method, the optimized query may fail.
+    /// Since it's quite rare case, we just ignore this possibility.
```

Comment:
I'm not sure if it's ok to leave this race as is. But there is no easy solution to avoid it now. Howerver, it could be fixed after introducing storage snapshot in #23932. In snapshot we can save all tables, from which `Merge` table will read, at the moment of start of  analysis of query.

Reply:
It's ok to have this race.

Reply:
I think we should tell that we support prewhere if at least one table support it.
Or something like merge(MergeTree, LOG) would not work.
For those which don't, we can add filtration step separately afterwards.